llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp

```diff
@@ -2041,6 +2041,28 @@ void CombinerHelper::applyCombineMulToShl(MachineInstr &MI,
   Observer.changedInstr(MI);
 }
 
+bool CombinerHelper::matchCombineSubToAdd(MachineInstr &MI,
+                                          BuildFnTy &MatchInfo) {
+  GSub &Sub = cast<GSub>(MI);
+
+  LLT Ty = MRI.getType(Sub.getReg(0));
+
+  if (!isLegalOrBeforeLegalizer({TargetOpcode::G_ADD, {Ty}}))
+    return false;
```

Comment:
Also should check isConstantLegalOrBeforeLegalizer here

Reply:
Although actually the legality rule doesn't consider the value now, and a typed constant already existed. We could revisit this if we had the value in the legalizer rule

Reply:
Although if you custom lowered the constant, it would still work to block it. So I guess it still sort of makes sense to check

Reply:
Done